### PR TITLE
Initial version of the code-signing dataplane service

### LINKED ISSUE #296: [PRSS - AZURE CODE SIGNING - SIGN - Azure Code Signing] API Review
New API Review meeting has been requested.

**Service Name**: PRSS - AZURE CODE SIGNING - SIGN - Azure Code Signing
**Review Created By**: LaDonna Quinn
**Review Date**: 11/16/2023 08:00 AM PT
**Onboarding Record**: https://dev.azure.com/azure-sdk/Release/_workitems/edit/19447
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/22310
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/9d71b54a-af65-4680-8f96-4671f6cff4e6?view=edit&event=e0fef6f8-6c7a-ee11-8179-000d3a5651e3)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### LINKED ISSUE #297: [PRSS - AZURE CODE SIGNING - SIGN - Azure Code Signing] API Review
New API Review meeting has been requested.

**Service Name**: PRSS - AZURE CODE SIGNING - SIGN - Azure Code Signing
**Review Created By**: LaDonna Quinn
**Review Date**: 11/16/2023 01:05 PM PT
**Onboarding Record**: https://dev.azure.com/azure-sdk/Release/_workitems/edit/19447
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/22310
**Hero Scenarios Link**: [here](https://microsoft-my.sharepoint-df.com/:w:/r/personal/rakiasegev_microsoft_com/_layouts/15/guestaccess.aspx?e=zma7Ey&share=EdK1ryeZ1LNAqwHmrcrQegoB2EZk552BVAo1zjnNiYUBHw&isSPOFile=1)
**Core Concepts Doc Link**: [here](https://microsoft.sharepoint.com/:p:/t/PRSS3rdPartySigningService/IQFa1SPxJWauTb217dmgwGW1AQOUmb8KIPPxVkEK3oDjzdU?e=rVvaDO)

**Description**: 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/9d71b54a-af65-4680-8f96-4671f6cff4e6?view=edit&event=b2f705e0-bc7c-ee11-8179-000d3a56548f)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### LINKED ISSUE #23716: [PRSS - AZURE CODE SIGNING - SIGN - Azure Code Signing] API Review
New API Review meeting has been requested.

**Service Name**: PRSS - AZURE CODE SIGNING - SIGN - Azure Code Signing
**Review Created By**: Jaxel Rojas Lopez @Jaxelr 
**Review Date**: 05/23/2023 01:00 PM PT
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/22310
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Scheduling API Review meeting for Dataplane service of Azure Code Signing service. 

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=596c0077-89e3-ed11-a7c7-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by Jaxelr**
# Data Plane API - Pull Request
Initial onboarding definition for the dataplane service of the Azure Code Signing service.
See: https://github.com/Azure/cadl-azure/issues/1999
cc: @DawnWang17 

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

fix https://github.com/Azure/azure-rest-api-specs/issues/23716
fix https://github.com/azure-sdk/TestingForEngagementExperience/issues/296
fix https://github.com/azure-sdk/TestingForEngagementExperience/issues/297

### CODE REVIEW

File: specification/codesigning/Microsoft.CodeSigning/main.cadl [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+#suppress "@azure-tools/cadl-azure-core/use-standard-operations" "This is a custom operation status endpoint."
+interface CertificateProfileOperations {
+  @summary("Gets the status of a sign operation.")
+  @doc("Gets the status of a sign operation.")
+  @route("/codesigningaccounts/{codeSigningAccountName}/certificateprofiles/{certificateProfileName}/sign/{operationId}")
+  getSignOperation is Foundations.GetOperationStatus<CodeSignStatusParams,OperationStatus>;
+
+  @summary("Initiates a sign operation.")
+  @doc("Initiates a sign operation.")
+  @pollingOperation("GetSignOperation")
```
[the remark this answers is not in the document]
Comment (Jaxelr) [reply specification/codesigning/Microsoft.CodeSigning/main.cadl]: Thanks, this is fixed on commit 384fd05

**Comment by Jaxelr** [comment]
@msftbot dont close, still had work scheduled.

**Review by mikekistler — CHANGES_REQUESTED** [review]
This looks very good but there are few minor problems. Most should be easy to fix.

File: specification/codesigning/Microsoft.CodeSigning/main.tsp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+@doc("Azure CodeSigning is a service that provides managed code signing for all.")
+namespace Azure.CodeSigning;
+
+enum Versions {
+  @useDependency(Azure.Core.Versions.v1_0_Preview_2)
+  v2022_06_15_preview: "2022-06-15-preview",
+}
+
+interface CertificateProfileOperations {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is a custom operation status endpoint."
```
[the remark this answers is not in the document]
Comment (Jaxelr) [reply specification/codesigning/Microsoft.CodeSigning/main.tsp:39]: The reason was that I needed a custom route with this path:

/codesigningaccounts/{codeSigningAccountName}/certificateprofiles/{certificateProfileName}/**sign**/{operationId}

and the GetResourceOperationStatus appends an operations path inside the path:

"/codesigningaccounts/{codeSigningAccountName}/certificateprofiles/{certificateProfileName}/**operations**/{operationId}"

File: specification/codesigning/Microsoft.CodeSigning/main.tsp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    "./examples/2022-06-15-preview/CertificateProfile_GetSignEku.json",
+    "GetSignEku"
+  )
+  GetSignEku is ResourceList<Eku>;
+
+  @summary("Initiates a sign operation.")
+  @doc("Initiates a sign operation.")
+  @example("./examples/2022-06-15-preview/CertificateProfile_Sign.json", "Sign")
+  @pollingOperation(CertificateProfileOperations.GetSignOperation)
+  Sign is ResourceAction<CertificateProfile, SignRequest, OperationStatus>;
```
[the remark this answers is not in the document]
Comment (Jaxelr) [reply specification/codesigning/Microsoft.CodeSigning/main.tsp]: I agree. Committed the modification.

**Review by mikekistler — APPROVED** [review]
Thanks for the fixes! This is looking really good now. I think it is ready to bring to the API Stewardship review whenever you are ready.

File: specification/codesigning/Microsoft.CodeSigning/tspconfig.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+parameters:
+  service-directory-name:
+    default: "codesigning"
+emit:
+  - "@azure-tools/typespec-autorest"
+options:
+  "@azure-tools/typespec-autorest":
```
Comment (weshaggard) [inline specification/codesigning/Microsoft.CodeSigning/tspconfig.yaml:7]: I assume you want to generate SDK's and if so, you will want add the other sdk configurations. See our sample at https://github.com/Azure/azure-rest-api-specs/blob/main/specification/contosowidgetmanager/Contoso.WidgetManager/tspconfig.yaml.

File: specification/codesigning/Microsoft.CodeSigning/package.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
```
[the remark this answers is not in the document]
Comment (weshaggard) [reply specification/codesigning/Microsoft.CodeSigning/package.json:1]: You might also need to update your branch with the latest changes in main to pull in the newer version of the package.json in the root of the repo.

File: specification/codesigning/Microsoft.CodeSigning/main.tsp [hunk 6]
```diff
@@ -0,0 +1,221 @@
+import "@typespec/rest";
```
Comment (weshaggard) [inline specification/codesigning/Microsoft.CodeSigning/main.tsp:1]: Let please remove "Microsoft." from the path as we try to keep those redundant names out of the paths. See https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/typespec-structure-guidelines.md for some of the structure guidelines.

File: specification/codesigning/Microsoft.CodeSigning/package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jaxelr) [reply specification/codesigning/Microsoft.CodeSigning/package.json]: This is done, thanks!

File: specification/codesigning/Microsoft.CodeSigning/tspconfig.yaml (hunk 4 above)
Comment (Jaxelr) [reply specification/codesigning/Microsoft.CodeSigning/tspconfig.yaml]: I used the sample for now to add those emitters. Thanks

File: specification/codesigning/Microsoft.CodeSigning/main.tsp (hunk 6 above)
Comment (Jaxelr) [reply specification/codesigning/Microsoft.CodeSigning/main.tsp]: Thank you, addressed feedback.

File: specification/codesigning/CodeSigning/main.tsp [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+@doc("Azure CodeSigning is a service that provides managed code signing for all.")
+namespace Azure.CodeSigning;
+
+enum Versions {
+  @useDependency(Azure.Core.Versions.v1_0_Preview_2)
+  v2022_06_15_preview: "2022-06-15-preview",
+}
+
+interface CertificateProfileOperations {
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is a custom operation status endpoint."
```
Comment (mikekistler) [inline specification/codesigning/CodeSigning/main.tsp:42]: Why can you not use an operation template from Azure Core or Foundations? It looks like Foundations.GetOperationStatus should work.

File: specification/codesigning/CodeSigning/main.tsp [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    "./examples/CertificateProfile_GetSignOperation.json",
+    "GetSignOperation"
+  )
+  @route("/codesigningaccounts/{codeSigningAccountName}/certificateprofiles/{certificateProfileName}/sign/{operationId}")
+  getSignOperation is Foundations.GetOperationStatus<
+    CodeSignStatus,
+    OperationStatus
+  >;
+
+  #suppress "@azure-tools/typespec-azure-core/use-standard-operations" "This is a custom operation status endpoint."
```
Comment (mikekistler) [inline specification/codesigning/CodeSigning/main.tsp:55]: The explanation for this suppression looks wrong -- looks like it was just copied from above. Please consider using an operation template or provide a better explanation here.

If you do not change to use an operation template, you must add an api-version parameter and a default (error) response.

**Review by mikekistler — CHANGES_REQUESTED** [review]
Please use operation templates from Azure Core wherever possible. If not possible, please provide a clear explanation.

**Review by markweitzel — APPROVED** [review]
Per #23716 -- API Stewardship Board approved for Preview

File: specification/codesigning/CodeSigning/main.tsp (hunk 7 above)
Comment (mikekistler) [reply specification/codesigning/CodeSigning/main.tsp:42]: Sorry ... I see now that you are using GetOperationStatus. Perhaps we need to promote this into Azure Core proper.

File: specification/codesigning/CodeSigning/main.tsp (hunk 8 above)
Comment (Jaxelr) [reply specification/codesigning/CodeSigning/main.tsp:55]: This will be resolved using the suggestion below.

File: specification/codesigning/CodeSigning/main.tsp (hunk 7 above)
Comment (Jaxelr) [reply specification/codesigning/CodeSigning/main.tsp:42]: My assumption is that there's no action needed from my end on this finding, as of now. I'm marking as resolved, but feel free to unresolve otherwise.

**Comment by Jaxelr** [comment]
@msftbot dont close, I've been on holidays and will pick up this work on this sprint.

File: custom-words.txt [hunk 9]
```diff
@@ -493,9 +494,8 @@ codegen
 codegeneration
 codemirror
 codepackage
-codespace
-codespaces
-coffeelake
+codesigning
+Coffeelake
```
Comment (weshaggard) [inline custom-words.txt:498]: Why did you remove the codespace entries? Also please use lowercase.

Comment (Jaxelr) [reply custom-words.txt:498]: I must've messed up the rebase. Thanks for the catch, this is fixed.

**Review by mikekistler — CHANGES_REQUESTED** [review]
One more small suggested change.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

**Comment by Jaxelr** [comment]
@weshaggard / @lmazuel anything else missing from my end?

File: specification/codesigning/CodeSigning/main.tsp [hunk 10]
```diff
@@ -0,0 +1,207 @@
+import "@typespec/rest";
+import "@typespec/versioning";
+import "@azure-tools/typespec-azure-core";
+import "@azure-tools/typespec-autorest";
```
Comment (mikeharder) [inline specification/codesigning/CodeSigning/main.tsp:4]:
```suggestion
```

Your spec should not requires this import.

**Review by mikeharder — DISMISSED** [review]
TypeSpec Validation is failing

File: specification/codesigning/CodeSigning/main.tsp (hunk 10 above)
Comment (Jaxelr) [reply specification/codesigning/CodeSigning/main.tsp]: Fixed. Thanks

**Comment by lmazuel** [comment]
Force merging the PR, as the broken CI are caused by known bugs:
- https://github.com/Azure/typespec-azure/issues/1674
- https://github.com/Azure/typespec-azure/issues/3349

**Comment by weidongxu-microsoft** [comment]
Note that the example of the LRO uses `azure-async-operation`. Therefore, either the example is wrong, or the LRO is wrong.

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/codesigning/data-plane/Azure.CodeSigning/preview/2023-06-15-preview/examples/CertificateProfileOperations_Sign.json#L25C10-L25C31

**Comment by Jaxelr** [comment]
I can submit a PR to fix it, thanks for catching this up.

**Comment by weidongxu-microsoft** [comment]
Thanks @Jaxelr 

If possible, would you also give us an example of the last poll (when success)?

As LongRunningResourceAction, the final poll should be as shape
```
{
  "id": "##",
  "status": "succeeded",
  "result": {
    ...
  }
}
```

**Comment by Jaxelr** [comment]
I can definitely do that, Thanks.

**Comment by weidongxu-microsoft** [comment]
Thanks. Related doc on LRO of LongRunningResourceAction
https://github.com/microsoft/api-guidelines/blob/vNext/azure/ConsiderationsForServiceDesign.md#long-running-action-operations
